From 5422569d6a84d4d358d4ef9eb4fd128a3b0f85d2 Mon Sep 17 00:00:00 2001
From: Michael Niedermayer <michael@niedermayer.cc>
Date: Tue, 13 Oct 2026 05:52:20 +0200
Subject: [PATCH] avformat/aviobuf: retry the read after AVERROR(EAGAIN)

Fixes: 3Fn44aM2tqIx
Fixes: infinite loop
Infinite loop Replicated through UnModified FFmpeg with a nonblocking pipe and timeout
Found during the review of https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/24909

(cherry picked from commit 5422569d6a84d4d358d4ef9eb4fd128a3b0f85d2)
---
 libavformat/aviobuf.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/libavformat/aviobuf.c b/libavformat/aviobuf.c
index edbbc5ffd67..43431381b9b 100644
--- a/libavformat/aviobuf.c
+++ b/libavformat/aviobuf.c
@@ -520,6 +520,11 @@ static void fill_buffer(AVIOContext *s)
                           s->buf_end : s->buffer;
     int len             = s->buffer_size - (dst - s->buffer);
 
+    if (s->error == AVERROR(EAGAIN)) {
+        s->error       = 0;
+        s->eof_reached = 0;
+    }
+
     /* can't fill the buffer without read_packet, just set EOF if appropriate */
     if (!s->read_packet && s->buf_ptr >= s->buf_end)
         s->eof_reached = 1;
@@ -622,6 +627,10 @@ int avio_read(AVIOContext *s, unsigned char *buf, int size)
         if (len == 0 || s->write_flag) {
             if((s->direct || size > s->buffer_size) && !s->update_checksum && s->read_packet) {
                 // bypass the buffer and read data directly into buf
+                if (s->error == AVERROR(EAGAIN)) {
+                    s->error       = 0;
+                    s->eof_reached = 0;
+                }
                 len = read_packet_wrapper(s, buf, size);
                 if (len == AVERROR_EOF) {
                     /* do not modify buffer if EOF reached so that a seek back can
-- 
2.53.0

